[RFC 0101] Nix formatting (#101)
NixOS/rfcs: 0101-nix formatting.md

## Future work

[future]: #future-work

- TODO General style guidelines beyond AST reformatting

## Appendix A: Formatting style guidelines

**Nomenclature:**
Brackets: `[]`
Braces: `{}`
Parentheses: `()`

*These are not strict rules but guidelines to help create more consistent formatting rules.
Nevertheless, deviations should be documented and explained.*

- When deciding between two *equally good* options, currently prevalent formatting style in `nixpkgs` should be followed.
- On the top level (indentation zero), special rules may apply.
- Two spaces are used for each indentation level, there is no vertical alignment
- Consistency across language features is important. Syntax that behaves similarly should be formatted similarly.
- The Nix formatting style is space heavy. Where there can be a space, there should be a space.
  - Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.
  - `[]` is written as `[ ]`, same for `{ }`.
  - Exception: Parentheses are written *without* a space on the inside
- Expressions are either written on one line, or maximally spread out across lines, with no in-between (e.g. grouping).

>   - Multi-line statements increment the indentation level in the statement's "body".

Reviewer: The specific rules below carve out one exception to this bullet point for top-level function bodies. I'd like to propose a few more (not limited to the top level):

```nix
assert ...;
big multi-line body {
  goes = here;
}
```

```nix
with ...;
big multi-line body {
  goes = here;
}
```

```nix
let
  x = ...;
in
big multi-line body {
  using = x;
}
```

These exceptions are motivated by my expectation that adding/removing assertions, factoring/inlining a `with`, and eliminating/inlining common subexpressions with `let` are all common code transformations that typically change only a few lines in the body (or zero, in the case of `assert`), and diffs would be less focused if performing these transformations required reindenting the entire body.

If this style guide doesn't forbid indentation in these bodies, I would at least want it not to be required.

Reviewer 2: please, unneeded indentation is always so unpleasing
like if the last `}` ends on indentation level 1 instead of 0
alejandra does this

---
feature: nix_formatting
start-date: 2021-08-17
author: Raphael Megzari
co-authors: (find a buddy later to help out with the RFC)

> shepherd-team: Timothy DeHerrera (nrdxp), 0x4A6F, Silvan Mosberger (infinisil)

Reviewer 3: ```suggestion
shepherd-team: Timothy DeHerrera (nrdxp), 0x4A6F, Silvan Mosberger (infinisil), piegames
```
I'm assuming that piegames has agreed to shepherd, based on their participation in the meeting and that this was one of the outcomes listed in the notes :)

> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 4: Please update the shepherd team here.

Author: added.

related-issues: (will contain links to implementation PRs)
---

# Summary

[summary]: #summary

Decide on a recommended automated formatter for nix files in nixpkgs.

# Motivation

[motivation]: #motivation

Prevent debate around how things should be formatted.

# Detailed design

[design]: #detailed-design

The implementation of this RFC should include several parts

> - Agree on an automated formatter. [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) is proposed here.

Reviewer 5: That formatter would need to become a channel blocker for the `nixpkgs-*` channels (like e. g. `nixpkgs-review` is).

Reviewer 6: I'd like to propose [Alejandra](https://github.com/kamadorueda/alejandra) to be taken into consideration as well.

> - If adopted, agree on a schedule to format nixpkgs codebase. Should it be done at once, or on a per package basis.

Reviewer 7: If nixpkgs-fmt changes the content of values, I would consider this as a bug. Unfortunately due to how the internal data structures that rnix-parser is using, deal with multi-line strings and comments has been particularly challenging.

If somebody could extract the repro and post an issue to nixpkgs-fmt that would be great.

- Potentially agree on a hook to enforce formatting

> 

Reviewer 8: I think a few formatting choices still need to be discussed. One of the largest issues is probably where the trailing commands in attrsets should be https://github.com/nix-community/nixpkgs-fmt/issues/248. The main arguments have probably already been said in that issue - I don't know if we should continue the discussion here or there.

Reviewer 9: I left a comment in the issue

> - Potentially agree on a hook to enforce formatting

Reviewer 7: This could be implemented as a github action pretty easily. `$(nix-build -A nixpgks-fmt)/bin/nixpkgs-fmt .`

Reviewer 10: There is also [treefmt](https://github.com/numtide/treefmt), which might be used to exclude specific path to address problematic files.
Here is a test run of treefmt on [nixpkgs](https://github.com/0x4A6F/nixpkgs/tree/master-fmt).

List to consider for exclusion:
- [rPackages](https://github.com/NixOS/rfcs/pull/101#issuecomment-929717228)

# Examples and Interactions

[examples-and-interactions]: #examples-and-interactions

> This section is not needed for this RFC.

Reviewer 8: I think if adding an exact specification of how files are formatted is too much work a few representative samples should at least be added here. I think that would be a fair compromise between effort and at least having some reference. Other than that the formatter at a specified commit is probably the ground truth.

# Drawbacks

[drawbacks]: #drawbacks

> - Having a commit that changes the formatting, can make git blame harder to use. It will need `--ignore-rev` flag.

Reviewer 11: Made a pr for a `.git-blame-ignore-revs` file
https://github.com/NixOS/nixpkgs/pull/162430

According to a github employee they're currently working on the web blame
> Update: implementation of this feature is in progress
> **6 days ago**

https://github.com/github/feedback/discussions/5033#discussioncomment-2237184

> - Every formatter will have bugs. An example of a bug for nixpkgs-fmt can be found [here](https://github.com/NixOS/nixpkgs/pull/129392)

Reviewer 12: [Alejandra](https://github.com/kamadorueda/alejandra) gets multi-line strings done perfectly, plus only 94 out of 34079 derivations would need rebuilding (and for reasons that can be happily ignored, see footnote on the readme)

# Alternatives

[alternatives]: #alternatives

> - Alternative formatter [nixfmt](https://github.com/serokell/nixfmt)

Reviewer 13: Enforces short line lengths which is not ideal with src in fetchers.

Reviewer 14: If single string literal is too long, it is kept, so I see little problem with it. Can you provide example when it generates outright ugly results?

> - Keep the status quo of not having an official formatter. The danger is that this creates discord within the nix community. On top of fragmented the community it can generate lengthy discussions (which do not advance the eco-system).

Reviewer 15: There is also an alternative of explicitly discouraging any comments on formatting not directly based on coding style recommendations spelled out in the Nixpkgs manual.

Reviewer 14: I'd say that it would be step in wrong direction. As far as I remember, nothing in nixpkgs manual says that your expression should contain newlines at all.

Reviewer 15: Well, discouraging is not the same as strictly forbidding, if you in good faith consider something an _exceptionally_ unfortunate formatting decision, you could comment about that… with a link to a PR to the manual, obviously.

Reviewer 16: Where would you draw the line? It should be explicit, otherwise one reviewer thinks one part is exceptionally unfortunate and the next thinks another part is exceptionally unfortunate.

# Unresolved questions

[unresolved]: #unresolved-questions

- Not sure how much work there is left on nixpkgs-fmt before most people would consider it ok to use. Not even sure how much it is actually used.

> - Are there situation where automated formatting is worse than manual formatting ? Do we want to have exceptions to automated formatting ?

Reviewer 15: I have to admit that updating code generation by putting the generator call inside a script and calling `nixpkgs-fmt` afterwards sounds like a nonzero but manageable amount of work.

On the other hand, maintaining generators to always be formatter-clean would create unncessary pushback to fixing some weird corner cases.

> - Are there situation where automated formatting is worse than manual formatting ? Do we want to have exceptions to automated formatting ?

Reviewer 17: IMO just fixup the generated files after they have been generated. Either the tool itself can do that, your pre-commit hooks, or the update script. Either of these works. If we enforce formatting of files then almost no file should be excluded. Especially not generated files as being able to read them is important. There is no excuse to treat those a black boxes that shouldn't be looked at.

> - Are there situation where automated formatting is worse than manual formatting ? Do we want to have exceptions to automated formatting ?

Reviewer 13: ```suggestion
- Are there situation where automated formatting is worse than manual formatting? Do we want to have exceptions to automated formatting?
```

> - Are there situation where automated formatting is worse than manual formatting ? Do we want to have exceptions to automated formatting ?

Reviewer 18: The original tool should be fixed to always comply, to avoid churn. Or we marke the file "no format". Or we do https://github.com/NixOS/rfcs/pull/109/

### Incremental migration

A more gradual approach, where only changed files get formatted for some period of time to make a more graceful transition has been rejected for practical concerns. Among other things, such an approach would not improve the amount of merge conflicts, and increase the workload for contributors during that time significantly.

## Unresolved questions

[unresolved]: #unresolved-questions

- Are there situation where automated formatting is worse than manual formatting? Do we want to have exceptions to automated formatting?

## Future work

[future]: #future-work

- TODO General style guidelines beyond AST reformatting

## Appendix A: Formatting style guidelines

**Nomenclature:**
Brackets: `[]`
Braces: `{}`
Parentheses: `()`

*These are not strict rules but guidelines to help create more consistent formatting rules.
Nevertheless, deviations should be documented and explained.*

- When deciding between two *equally good* options, currently prevalent formatting style in `nixpkgs` should be followed.
- On the top level (indentation zero), special rules may apply.
  - TODO: Is this still relevant?

> - Two spaces are used for each indentation level, there is no vertical alignment

Reviewer 19: ```suggestion
- One tab is used for each indentation level, there is no vertical alignment
```

Tabs have real world benefits for making the code more accessible for the visually impaired. 

I already outlined this a bit [here](https://github.com/NixOS/rfcs/pull/101#issuecomment-1707814452) .
I believe this is a point we should at least discuss.

Reviewer 20: I am personally a huge fan of tabs for indentation too, and when I started Nix I used tabs for my configuration at first, and even kept it that way for years until [only nine months ago](https://git.darmstadt.ccc.de/piegames/home-config/-/commit/27ace65e09aa75d29dc2fe3c62678ff95ead5916). The big problem with tabs is that they are not supported in Nix strings in the same way spaces are. Using tabs for indentation means you must deal with that fact in some way.

In my home configuration, I created a small [`stripTabs`](https://git.darmstadt.ccc.de/piegames/home-config/-/commit/bfd7905d08b89f96b7a8cfcbf30eec8cfb83a132) helper function which emulates the "smart" behavior of Nix strings. This has the disadvantage that one must wrap every single multi line string in a function call. This is easy to forget (for example when a single line string becomes multi line), easy to miss and can cause really confusing behavior. (Most of the time things will work just fine and just look a bit weird should you happen to look at the generated output, until they don't.)

> as whitespace on strings have semantics that cannot be understood by a nix formatter

This is wrong. *Because* of the semantics of strings in Nix, indentation rules also apply *within* a string. The formatter is allowed to change characters within the string as long as it still evaluates to the same value. This also applies to interpolation and escaping by the way. Those strings only work with space indentation, thus unfortunately forcing that onto the rest of the language.

> literally zero cost

Please show me a zero-cost solution for how to use Tabs with Nix, preferably in a non-trivial production code base. I've shown you mine, and it was not zero cost at all. (I do know of some other ways to hack around this, but they all have their own significant downsides.)

-------

I don't see a way around using spaces for indentation in the current state of the Nix language, but I'd love to see the language fixed one day and then I'd be happy to move everything over to tabs. Unfortunately, this will likely require backwards-incompatible changes and thus depend on something like [RFC 137](https://github.com/NixOS/rfcs/pull/137)

Reviewer 21: See also https://github.com/NixOS/nix/issues/7834

Reviewer 22: Tabs over spaces forever tbh, way more easier to do and just easier to deal with imo.

Reviewer 2: > I'd love to see the language fixed one day and then I'd be happy to move everything over to tabs. 

I don't think we can move over to tabs after this is finished, we would have to literally change every line of code in Nixpkgs again.

I'd still be in favor of tabs, but:
To use tabs as indentation, we have to either wait with this RFC or use tabs for everything outside multiline strings until we have support for tabs in multiline strings.
Since we don't actually know if tab support will ever happen, betting on being able to make it more consistent later is not preferable in my opinion.
Or we just accept the inconsistency (spaces in multiline strings and tabs everywhere else) and just hope we can make it more consistent without betting on it.

Reviewer 19: > Please show me a zero-cost solution for how to use Tabs with Nix

Just to clarify, my argument was not regarding the semantics that tabs vs strings have regarding the nix codebase. My point is rather that tabs have little to no downside to the average human code reader / writer. I agree I could have phrased that better.

Reviewer 23: Much of this talk seems to be off topic for this RFC, no?

Reviewer 24: Preferences aside, I don't think using tabs in nixpkgs is feasible, since there are only 2 options I see

- Use spaces as a workaround whenever idented strings are used - This makes the code only look right at one tab width, which defeats the purpose of using tabs in the first place
- Fix Nix and raise the minver - Keep in mind the nixpkgs's minver is still at 2.3, adopting tabs would require the fix to be implemented in Nix, and we would need to raise the minver to a version that supports flakes, which would open a new can of worms that will probably take a very long time to reach an agreement.

Reviewer 19: > Preferences aside

I think that's part of the problem right there. This discussion is not about preference. For someone in a wheelchair, using a ramp instead of the stairs is not a matter of preference.
This argument is about whether or not we want the nixpkgs code to be more accessible and this code formatting standard should reflect that decision.

Reviewer 24: My point is not that we should not switch to tabs, but that we should not wait for the fix in nix and a debate about minver to make the decision on this RFC. I think it's perfectly fine to list this in unresolved questions and move on with this RFC since this is not a regression (we are already enforcing 2-space indentation).

Reviewer 20: > I don't think we can move over to tabs after this is finished, we would have to literally change every line of code in Nixpkgs again.

Well yes. We are already going to touch every single code line now with this RFC, and we could always do another RFC for moving towards tabs and do it again. I'd be willing to do that part, but I won't be the one to fix the language, which is a prerequisite.

> To use tabs as indentation, we have to either wait with this RFC or use tabs for everything outside multiline strings until we have support for tabs in multiline strings.

This is a common suggestion, and additionally to the aforementioned problems it also does not interact well with nested string interpolations. Nix code can easily have three levels of interpolation nesting. Good luck properly dealing with that as a sighted user, not even speaking of how to deal with mixed tabs and spaces using a screen reader (it's possible, but a PITA).

Also, delaying this RFC even further for this reason is a non-starter IMO, especially since we are talking about a delay of probably at least several years. Formatting Nixpkgs is long overdue, don't let perfect be the enemy of the good here.

> My point is rather that tabs have little to no downside to the average human code reader / writer.

Generally speaking, yes. But not in Nix. Again, please show me a Nix code base that uses tabs for indentation without too much insanity. I am open to the idea, just very very skeptical because of my prior attempts.

Reviewer 19: > Formatting Nixpkgs is long overdue, don't let perfect be the enemy of the good here.

Totally agree on that. It's been one of my major pain points when getting code review for pull requests on nixpkgs.
We could throw `alejandra` onto the code base tomorrow and be done with good enough.
I seems rather this RFC has gone the down the perfect road long before I chimed in here so we might as well finish in style.

Reviewer 20: > We could throw alejandra onto the code base tomorrow and be done with good enough.

IMO, the current output from alejandra is "good enough". Personally, I define "good enough" as "better than the current state of Nixpkgs of average, and where better manual formatting could be achieved it wouldn't be worth the time and effort". For me, the nixfmt branch has reached that goal some time ago, and since then further improvements have brought it into a state where it is hopefully good enough for more critical people. But this is off-topic for this thread.

Reviewer: > What I meant is, that the community is not likely to accept another big Nixpkgs refactor just for the sake of changing indentation to tabs. But this is probably the best option.

As a random community contributor (who is weakly pro-spaces—I won't prosecute the argument here, but it's more than just ‘I like it better’), I would be much more indulgent toward a future big refactor switching to tab indentation than I would toward just about any future large-scale tweak to some other formatting decision. There are technical pros and cons to the tabs-versus-spaces conversation; switching initial whitespace from one to the other is easy to automate even without building and running the latest iteration of the blessed formatter; and unlike most formatting decisions, it's entirely orthogonal to any other change that is likely to be in flight.

Reviewer 25: To be honest I feel like the team responsible for formatting should just have carte blanche to alter formatting rules and format the tree accordingly, within reason (e.g. please batch it up so you're not doing it more than once per every couple releases at least!); the process of this RFC has demonstrated pretty well that the RFC format is not conducive to bikeshedding formatting decisions. With `.git-blame-ignore-revs`, the main issue is merge conflicts in PRs, which is admittedly not nothing, but is also a fact of life when contributing to Nixpkgs in general. So I would personally be content to see a change like this done without excessive amounts of red tape.

Reviewer 19: In my opinion, any half way sensible formatter / linter is better than no formatter. The rules themselves matter very little, the important part is that rules exist, they are transparent an enforcible and contributors can get it right the first time and there is no nitpicking over whitespace in pull requests.

I agree with @emilazy , just give someone authority over this and get it done. It's not a topic that will be solved in open public conversation. For reasons that are beyond me, people have extremely strong opinions on the topic that can rarely be unified to a degree that really satisfies everyone.

inherit foo bar;
inherit (attrs) foo bar;
inherit
  foo
  bar
  ;
inherit (attrs)
  foo
  bar
  ;
```

- Inherit expressions can have their symbols either all on the same line as the `inherit`, or one line per element each.
- The `(attrs)` is always on the same line as the `inherit` statement. There are no spaces on the inside of its parentheses.

❌ Bad:
```nix
TODO
inherit (attrs) foo
  bar
  ;
```

❌ Bad:
```nix
inherit
  (attrs)
  foo
  bar
  ;
```

✅ Good:

```nix
{
  inherit
    (mkFruit {
      a = "apple";
      b = "banana";
      c = "cherry";
      d = "date";
      e = "elderberry";
    })
    feijoa
    grape
    honeydew
    ;
}
```

- Technically this is inconsistent because the attribute set is sometimes on the same line as the `inherit` and other times not. However, this is a really niche use case and generally you won't need to write such code anyways so we are willing to sacrifice consistency for a better default.

### If-Then-Else

✅ Good:

```nix
if predicate then
  foo
else
  bar
```

- "Pulling up" the `then` keyword is consistent with other similar syntax constructs
- The bodies are always indented

❌ Bad (old):

```nix
if predicate
then foo
else bar
```

> - TODO explain why bad

Reviewer 26: I think a good reasoning would include multi-line expressions inside the then and else blocks:

```nix
if something <= 4.0
then
   {
     foo = 10;
   }
else if something <= 6.0
then
   [
     10
     20
   ]
else if something <= 7.0
then
   let
     …
   in
     …
```

Is overly long and has no advantage.

```nix
if something <= 4.0
then {
     foo = 10;
   }
else if something <= 6.0
then [
     10
     20
   ]
else if something <= 7.0
then let
     …
   in
     …
```

Is just as long as having `then` in line, but influences all the other formatting rules and it’s harder to tell at a glance where the blocks start and what they are.

`then` is really just a keyword for the parser, for the human reading it’s just noise. `if` and `else` are the important parts of conditionals.

- ~~The keyword at the beginning of the line
  states clearly the meaning of the content that follows.~~
- ~~Produces a clean diff when you add more code.
  For example: adding content to the `else`
  only produces a diff in the `else`.~~

❌ Bad:

```nix
if predicate then foo else bar
```

> - One-liners are hard to understand,

Reviewer 26: I would argue against this in simple cases. Maybe there could be a limit for line length? If it is passed, the one-liner must be expanded.

```
env = {
  inherit version;
  release = if build.debug then 0 else 1;
  minify = if opts.minify then 1 else 0;
  codename = opts.name;
}
compileFlags =
  if build.debug then
    {
      wall = true;
      werror = true;
      optimize = 0;
    }
  else
    {
       ...
     }
  ;
```

I'm also not sure how to handle the conditional inside the assignment here. The hanging semicolon feels awkward.

Reviewer 26: Hmm but now we're absorbing braces again. Maybe that's ok? The draft right now specifically doesn't want to do this, and the reasoning is sound in my opinion. I get that not absorbing them indents the attributes another level and adds four additional lines in this case, but really, using the braces like you did doesn't make much sense in nix, as they don't delimit blocks and are thus not part of the control sequence.

But I do agree that we could just add the semicolon at the end. It's not that important semantically;

```nix
let
  compileFlags =
    if build.debug then
      {
        wall = true;
        werror = true;
        optimize = 0;
      }
    else
      {
        _ = "...";
      };
  env = {
    inherit version;
    release = if build.debug then 0 else 1;
    minify = if opts.minify then 1 else 0;
    codename = opts.name;
  };
in { }
```

You see that even without the semicolon, it's very clear that a new attribute starts.

❌ Bad:

```nix
if cond
then if
  looooooooooooooooooooooooooooooooooooooooooooooooooooong
then foo
else bar
else if cond
then foo
else bar
```

- It's complex to distinguish the parent `if-then-else`
  from the child `if-then-else`

❌ Bad:

```nix
if cond1 then [
  foo
] else if cond2 then
  bar
else if cond3 then {
  baz = "…";
} else if cond4 then let
  qux = "…";
in 
  qux
else
  true
```

- Inconsistent start of line
- if/else keywords not aligned: the first `else` is indented like if it was nested
- Easy to miss the opening brace/bracket, especially when the closing one is far below
- No clear distinction whether items are in a list or not

✅ Good:

```nix
if
  multiline
    && condition
then
  content
else if
  myPredicate [
    multi
    line
  ]
then
  other content
else
  more
```

- If the condition does not fit one line, expand with indentation, and put the `then` keyword on a new line
- Recommendation: avoid long conditions

### Lists

✅ Good:
```nix
[ foo bar baz ]
[
  foo
  bar
  baz
]
[ ]
```

- For consistency with other formatting rules, more spacing is preferred

❌ Bad:
```nix
[foo bar baz]
[]
```

### Function applications

✅ Good:
```nix
fun arg1 arg2 arg3

> callPackage ./some/path {

Reviewer 26: Why are curly braces of sets absorbed here but not in if else?

Reviewer 20: We special-cased the last argument of a function call, because it's extremely common

Reviewer 26: That's not an explanation. Having conditionals return one set or another set is also common and could just as well have inline braces:

``nix
if something <= 2.0 then
  if somethingElse then
    foo
  else
    bar
else if something <= 4.0 then {
  foo = 10;
} else if something <= 6.0 then [
  10
  20
]
```

Reviewer 26: Ah right but I see, the drawbacks of absorbing in conditionals are not relevant in function calls.

  foo = "bar";
} (
  if foo
  then bar
  else baz
)

if something <= 2.0
then
  if somethingElse
  then foo
  else bar
else if something <= 4.0
then
  {
    foo = 10;
  } // {
    bar = 20;
  }
else if something <= 6.0
then (
  if foo
  then bar
  else baz
)
else bar

> callPackage

Reviewer 26: Yeah this makes more sense, though it is a bit more verbose.

shepherd-team: 0x4A6F, Silvan Mosberger (infinisil), piegames, Thomas Bereknyei (tomberek)
shepherd-leader: Silvan Mosberger (infinisil)
related-issues: (will contain links to implementation PRs)
---

# Summary

[summary]: #summary

Decide on basic Nix style guidelines and pick a default code formatter for Nix code. 
Format Nixpkgs using that formatter and enforce it using CI.

## Motivation

[motivation]: #motivation

TODO: Currently, there is no authoritative style guide for Nix code in Nixpkgs.
The currenty style has evolved organically over time, leading to inconsistencies both across files and language features.

- We want to prevent debate around how things should be formatted.
- We want to make it as easy as possible for contributors (especially new ones) to make changes without having to worry about formatting.
- Conversely, reviewers should not be bothered with ill-formatted contributions.
- We want a unified Nix code style that's consistent with itself, easily readable, accessible and results in readable diffs.

## Detailed design

[design]: #detailed-design

There are four main parts to this RFC:
- Establish basic formatting guidelines

> - Pick a formatter implementation
> - Migrate Nixpkgs to the format and enforce it in CI

Reviewer 27: ```suggestion
```

I feel like these two shouldn't be part of the RFC but rather its implementation.

As @infinisil said, as long as the tool adheres to the style decided here, it doesn't matter which one you use. I'd just leave this open or at most give a recommendation based on experience in creating this RFC. Obviously said recommendation would be the most sensible to implement at first but if it later turns out that some other tool would be better suited, you could start using that without having to create a new RFC as it's an objective decision.

What really need to be decided through the RFC process are the style and rules as they're subjective and therefore need a community decision.
